File: ui/src/client/api/group.ts
// Copyright (c) 2023 Anass Bouassaba.
//
// Use of this software is governed by the Business Source License
// included in the file LICENSE in the root of this repository.
//
// As of the Change Date specified in that file, in accordance with
// the Business Source License, use of this software will be governed
// by the GNU Affero General Public License v3.0 only, included in the file
// AGPL-3.0-only in the root of this repository.
import useSWR, { SWRConfiguration } from 'swr'
import { PermissionType } from '@/client/api/permission'
import { apiFetcher } from '@/client/fetcher'
import { Organization } from './organization'

export enum GroupSortBy {
  Name = 'name',
  DateCreated = 'date_created',
  DateModified = 'date_modified',
}

export enum GroupSortOrder {
  Asc = 'asc',
  Desc = 'desc',
}

export type Group = {
  id: string
  name: string
  organization: Organization
  permission: PermissionType
  createTime: string
  updateTime?: string
}

export type GroupList = {
  data: Group[]
  totalPages: number
  totalElements: number
  page: number
  size: number
}

export type GroupListOptions = {
  query?: string
  organizationId?: string
  size?: number
  page?: number
  sortBy?: GroupSortBy
  sortOrder?: GroupSortOrder
}

export type GroupCreateOptions = {
  name: string
  image?: string
  organizationId: string
}

export type GroupPatchNameOptions = {
  name: string
}

export type GroupAddMemberOptions = {
  userId: string
}

export type GroupRemoveMemberOptions = {
  userId: string
}

type GroupListQueryParams = {
  page?: string
  size?: string
  sort_by?: string
  sort_order?: string
  query?: string
  organization_id?: string
}

export class GroupAPI {
  static create(options: GroupCreateOptions) {
    return apiFetcher({
      url: `/groups`,
      method: 'POST',
      body: JSON.stringify(options),
    }) as Promise<Group>
  }

  static patchName(id: string, options: GroupPatchNameOptions) {
    return apiFetcher({
      url: `/groups/${id}/name`,
      method: 'PATCH',
      body: JSON.stringify(options),
    }) as Promise<Group>
  }

  static useGet(id: string | null | undefined, swrOptions?: SWRConfiguration) {
    const url = `/groups/${id}`
    return useSWR<Group>(
      id ? url : null,
      () => apiFetcher({ url, method: 'GET' }) as Promise<Group>,
      swrOptions,
    )
  }

  static useList(options?: GroupListOptions, swrOptions?: SWRConfiguration) {
    const url = `/groups?${this.paramsFromListOptions(options)}`
    return useSWR<GroupList>(
      url,
      () => apiFetcher({ url, method: 'GET' }) as Promise<GroupList>,
      swrOptions,
    )
  }

  static paramsFromListOptions(options?: GroupListOptions): URLSearchParams {
    const params: GroupListQueryParams = {}
    if (options?.query) {
      params.query = encodeURIComponent(options.query.toString())
    }
    if (options?.organizationId) {
      params.organization_id = options.organizationId.toString()
    }
    if (options?.page) {
      params.page = options.page.toString()
    }
    if (options?.size) {
      params.size = options.size.toString()
    }
    if (options?.sortBy) {
      params.sort_by = options.sortBy.toString()
    }
    if (options?.sortOrder) {
      params.sort_order = options.sortOrder.toString()
    }
    return new URLSearchParams(params)
  }

  static delete(id: string) {
    return apiFetcher({
      url: `/groups/${id}`,
      method: 'DELETE',
    })
  }

  static addMember(id: string, options: GroupAddMemberOptions) {
    return apiFetcher({
      url: `/groups/${id}/members`,
      method: 'POST',
      body: JSON.stringify(options),
    })
  }

  static removeMember(id: string, options: GroupRemoveMemberOptions) {
    return apiFetcher({
      url: `/groups/${id}/members`,
      method: 'DELETE',
      body: JSON.stringify(options),
    })
  }
}

File: admin/api/database/group.py
# Copyright 2024 Piotr Łoboda.
#
# Use of this software is governed by the Business Source License
# included in the file licenses/BSL.txt.
#
# As of the Change Date specified in that file, in accordance with
# the Business Source License, use of this software will be governed
# by the GNU Affero General Public License v3.0 only, included in the file
# licenses/AGPL.txt.

from psycopg import DatabaseError

from ..dependencies import conn, parse_sql_update_query


# --- FETCH --- #
def fetch_group(_id: str):
    try:
        with conn.cursor() as curs:
            data = curs.execute(f'SELECT g.id as "group_id", g."name" as "group_name", g.create_time, g.update_time'
                                f'o.id as "org_id", o."name" as "org_name", o.create_time as "org_create_time", '
                                f'o.update_time as "org_update_time"'
                                f'FROM "group" g '
                                f'JOIN organization o ON g.organization_id = o.id '
                                f'WHERE g.id=\'{_id}\'').fetchone()

            return {'createTime': data.get('create_time'),
                    'id': data.get('group_id'),
                    'name': data.get('group_name'),
                    'organization': {
                        'id': data.get('org_id'),
                        'name': data.get('org_name'),
                        'createTime': data.get('org_create_time'),
                        'updateTime': data.get('org_update_time')
                    }
                    }
    except DatabaseError as error:
        raise error


def fetch_groups(page=1, size=10):
    try:
        with conn.cursor() as curs:
            data = curs.execute(f'SELECT g.id as "group_id", g."name" as "group_name", g.create_time '
                                f'as "group_create_time", g.update_time as "group_update_time", '
                                f'o.id as "org_id", o."name" as "org_name", o.create_time as "org_create_time", '
                                f'o.update_time as "org_update_time"'
                                f'FROM "group" g '
                                f'JOIN organization o ON g.organization_id = o.id '
                                f'ORDER BY g.create_time '
                                f'OFFSET {(page - 1) * size} '
                                f'LIMIT {size}').fetchall()

            count = curs.execute(f'SELECT count(1) FROM "group"').fetchone()

            return ({'id': d.get('group_id'),
                     'createTime': d.get('group_create_time'),
                     'updateTime': d.get('group_update_time'),
                     'name': d.get('group_name'),
                     'organization': {
                         'id': d.get('org_id'),
                         'name': d.get('org_name'),
                         'createTime': d.get('org_create_time'),
                         'updateTime': d.get('org_update_time'),
                     }
                     } for d in data), count['count']

    except DatabaseError as error:
        raise error


# --- UPDATE --- #
def update_group(data: dict):
    try:
        with conn.cursor() as curs:
            curs.execute(parse_sql_update_query('group', data))
    except DatabaseError as error:
        raise error

# --- CREATE --- #

# --- DELETE --- #
